bazel_tools/javadoc_library.bzl

```diff
@@ -0,0 +1,89 @@
+# Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+# This is a mixture of the javadoc_library rule in
+# https://github.com/google/bazel-common and the one in
+# https://github.com/stackb/rules_proto.
+
+# The main two factors for why we need this are
+
+# 1. We only have a source jar for java_proto_library.
+# The rules in bazel-common do not support this.
+
+# 2. We need to get the source jar via the JavaInfo provider.
+# The rules in rules_proto do not support this.
+
+# We probably want to clean this up and upstream it at some point.
+
+def _javadoc_library(ctx):
+    transitive_deps = []
+    for dep in ctx.attr.deps:
+        if JavaInfo in dep:
+            transitive_deps.append(dep[JavaInfo].transitive_deps)
+        elif hasattr(dep, "java"):
+            transitive_deps.append(dep.java.transitive_deps)
+
+    classpath = depset([], transitive = transitive_deps).to_list()
+
+    java_home = str(ctx.attr._jdk[java_common.JavaRuntimeInfo].java_home)
+
+    javadoc_command = [
+        java_home + "/bin/javadoc",
+        "-sourcepath .",
+        " ".join(ctx.attr.root_packages),
+        "-use",
+        "-subpackages",
+        ":".join(ctx.attr.root_packages),
+        "-encoding UTF8",
+        "-classpath",
+        ":".join([jar.path for jar in classpath]),
+        "-notimestamp",
+        "-d tmp",
+        "-Xdoclint:-missing",
+        "-quiet",
+    ]
+
+    if ctx.attr.doctitle:
+        javadoc_command.append('-doctitle "%s"' % ctx.attr.doctitle)
+
+    if ctx.attr.exclude_packages:
+        javadoc_command.append("-exclude %s" % ":".join(ctx.attr.exclude_packages))
+
+    for link in ctx.attr.external_javadoc_links:
+        javadoc_command.append("-linkoffline {0} {0}".format(link))
+
+    if ctx.attr.bottom_text:
+        javadoc_command.append("-bottom '%s'" % ctx.attr.bottom_text)
+
+    jar_command = "%s/bin/jar cf %s -C tmp ." % (java_home, ctx.outputs.jar.path)
```

Comment:
Same here, `jar.exe`?

Reply:
IIRC `jar cf` sets a timestamp which makes this indeterministic. It's probably better to use zipper instead.

Reply:
I’d like to not deviate further from upstream than we have to and this is how the upstream rules work. Given that nothing depends on this, this doesn’t seem like an issue for now and if it becomes an issue we should fix all javadocs (and source jars) at once.